Geometry Optimization Tutorials (13!)

psi-rking opened the pull request:
## Description
5 awesome geometry optimization tutorials!
All external optking dependencies removed, with helper functions streamlined and provided.

In the future, I will have students implement new methods in this way first.  I wish I had these a few year ago...  Would be nice to add P-RFO for transition states and possibly IRC (though that gets complicated) but no time for that right now.

## Changes
  - Added Tutorials/13 Geometry Optimization

## Status
[X] Click when ready for review-and-merge

loriab commented:
Neat, I want to read these!

If you would, please
* move your "Tutorials/12" to "Tutorials/13" (Andy's got 12)
* make a new file `test_TU_13.py` analogous to https://github.com/psi-rking/psi4numpy/blob/master/tests/test_TU_01.py that makes a test case for each of your ipynb's. If they need special software to run, you can include the decorators from [here](https://github.com/psi4/psi4numpy/blob/master/tests/addons.py#L34-L50) (or leave that part to me)

psi-rking commented:
13?  How come I have to be 13?   Ah well, I suppose nobody else will claim that one.

loriab commented:
You're welcome to choose among >=13 :-)

loriab commented:
Sorry to update your branch directly, @psi-rking. Not sure why it was letting me do that but not to do a PR to you. Anyway, this should help the tests.

loriab commented:
Ah, I take it pyoptking is needed? That falls under "special software to run" :-)

Will have to look more closely to answer your question about whether to bring it in.

psi-rking commented:
No trouble at all with you modifying the branch directly.  On the pyOptking
dependency, once I get through the initial set of tutorials, I think we
have 3 options:
a) require pyOptking imports
b) move anything needed into the psi4numpy repo, so it is locally
independent
c) embed all python needed into the ipynb files
Advantages and disadvantages.  I see c) getting too messy.  b) might be a
reasonable compromise.

On Tue, Nov 14, 2017 at 4:06 PM, Lori A. Burns <notifications@github.com>
wrote:

> Ah, I take it pyoptking is needed? That falls under "special software to
> run" :-)
>
> Will have to look more closely to answer your question about whether to
> bring it in.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/psi4/psi4numpy/pull/37#issuecomment-344414946>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADguTKvqYJCQH2GTI4imduh5lEvMXnM3ks5s2g7fgaJpZM4Qd2O2>
> .
>

loriab commented:
Yes, if it's mostly helper functions, I agree with (b), as is done [here](https://github.com/psi4/psi4numpy/tree/master/Tutorials/04_Density_Functional_Theory).

psi-rking commented:
OK, that will be my objective - to remove the external dependencies but to
keep the ipynb clean as possible.

On Tue, Nov 14, 2017 at 4:14 PM, Lori A. Burns <notifications@github.com>
wrote:

> Yes, if it's mostly helper functions, I agree with (b), as is done here
> <https://github.com/psi4/psi4numpy/tree/master/Tutorials/04_Density_Functional_Theory>
> .
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/psi4/psi4numpy/pull/37#issuecomment-344417024>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADguTF2GZ3_0ihcteG8Kbd9oUMap8pwJks5s2hC2gaJpZM4Qd2O2>
> .
>

psi-rking commented:
The tutorials are ready now.

dgasmith commented:
Cool! Can @dsirianni take a first pass at this?

dsirianni commented on Tutorials/13_Geometry-Optimization/opt_helper/physconst.py:
```diff
@@ -0,0 +1,3 @@
+pi = 3.14159265358979323846264338327950288
+bohr2angstroms = 0.52917720859
```
This file can safely be deprecated to avoid duplication, since each of these are available through either Psi4 or NumPy/Python `math`.  Pi is available (to the same `float64` precision as the above line) with both `np.pi` and `math.pi`, and the physical constants are available with `psi4.constants.bohr2angstroms` and `psi4.constants.hartree2aJ`.

psi-rking commented:
OK, removed physconst.py.   Using np.pi is no problem.

This commit does redundantly define the other two constants; for tutorial
purposes, does it matter?  The reason I wasn't using psi4.constants is that
1) the new pyOpkting doesn't require psi4 (and this code was copied and
modified from there); and 2) I don't want low-level files like stre.py to
depend upon psi4.

Anyway, I am considering this Tutorial helper code as standalone and
psi4-dependent, so we can deal with this any way that you would like.


On Wed, Jan 10, 2018 at 8:31 AM, Dominic A. Sirianni <
notifications@github.com> wrote:

> *@dsirianni* commented on this pull request.
> ------------------------------
>
> In Tutorials/13_Geometry-Optimization/opt_helper/physconst.py
> <https://github.com/psi4/psi4numpy/pull/37#discussion_r160692217>:
>
> > @@ -0,0 +1,3 @@
> +pi = 3.14159265358979323846264338327950288
> +bohr2angstroms = 0.52917720859
>
> This file can safely be deprecated to avoid duplication, since each of
> these are available through either Psi4 or NumPy/Python math. Pi is
> available (to the same float64 precision as the above line) with both
> np.pi and math.pi, and the physical constants are available with
> psi4.constants.bohr2angstroms and psi4.constants.hartree2aJ.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/psi4/psi4numpy/pull/37#pullrequestreview-87852531>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADguTG06O59oSm3cm1zZx3h_qmyMgUpdks5tJMmpgaJpZM4Qd2O2>
> .
>

dgasmith commented on Tutorials/13_Geometry-Optimization/13a_Internal-Coordinates-Bmatrix.ipynb:
```diff
@@ -0,0 +1,439 @@
+{
```
For the file, please do the following:

1) Group all import together at the top of a file in its own box.
2) Please do not use `v3d` code. `np.linalg.norm` works just as well.
3) `print()` does not have a space between the work and first parenthesis. Also every `print` statement need parenthesis to be Py3 comparable.
4) Please follow the PEP style guide. You can use `YAPF` to help you auto format your code.
5) I might avoid pulling from pubchem here, while cool those geometries can change due to how we match from that repo.
6) There is no need to use `fabs` unless you do not have floats.
7) Please use snake case for all functions and camel case for all classes via PEP.

psi-rking commented:
Thanks for the tip on fabs; I can't recall why I started using math's.

Good idea on the pubchem usage;  I removed it.

I've consolidated all the imports to one place in each tutorial file.

I forgot to fix those print statements; I had meant to do that and will do
it right now.

Agreed there is no need for v3d.norm.  In fact, it IS "np.linalg.norm()".
I'll phase that stuff out.

However, also in v3d I've got angle, torsion, and out-of-plan angle code
that has to handle special cases and exceptions in just the right way.

-Rollin

--
Dr. Rollin A. King
Professor and Chair of Chemistry
Bethel University
rking@bethel.edu

On Sat, Jan 13, 2018 at 4:14 PM, Daniel Smith <notifications@github.com>
wrote:

> *@dgasmith* approved this pull request.
>
> Looks good. My comments apply to all of your notebooks. I think its
> probably best to get this in soon and I can take a deeper look in situ.
> ------------------------------
>
> In Tutorials/13_Geometry-Optimization/13a_Internal-
> Coordinates-Bmatrix.ipynb
> <https://github.com/psi4/psi4numpy/pull/37#discussion_r161382720>:
>
> > @@ -0,0 +1,439 @@
> +{
>
> For the file, please do the following:
>
>    1. Group all import together at the top of a file in its own box.
>    2. Please do not use v3d code. np.linalg.norm works just as well.
>    3. print() does not have a space between the work and first
>    parenthesis. Also every print statement need parenthesis to be Py3
>    comparable.
>    4. Please follow the PEP style guide. You can use YAPF to help you
>    auto format your code.
>    5. I might avoid pulling from pubchem here, while cool those
>    geometries can change due to how we match from that repo.
>    6. There is no need to use fabs unless you do not have floats.
>    7. Please use snake case for all functions and camel case for all
>    classes via PEP.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/psi4/psi4numpy/pull/37#pullrequestreview-88668493>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADguTNjdIlD1X8rcXuPugdcgfG5w6yUhks5tKSrXgaJpZM4Qd2O2>
> .
>

loriab commented on Tutorials/13_Geometry-Optimization/opt_helper/v3d.py:
```diff
@@ -0,0 +1,171 @@
+def axpy(a, X, Y):
+    Z = np.zeros(Y.shape, float)
+    Z = a * X + Y
+    return Z
```
If `a` is scalar (usual daxpy), this is just `a * X + Y` in calling code, no need for fn.

loriab commented on Tutorials/13_Geometry-Optimization/opt_helper/oofp.py:
```diff
@@ -0,0 +1,135 @@
+        check, tau = v3d.oofp(geom[self.A], geom[self.B], geom[self.C], geom[self.D])
+        if not check:
+            raise optExceptions.ALG_FAIL(
+                "OOFP::compute.q: unable to compute out-of-plane value")
```
It would be more usual for the fns in v3d that hit some tolerance that currently results in a `check = False` to instead raise an exception right when the error occurs in v3d, rather than passing the flag back up to here to handle. That way the calling function could proceed assuredly. Is there a reason to defer error handling back to the calling fn, like the above?

loriab replied:
Got it. I'm far from demanding it be changed. The pattern I notice is that right now, some of the fns are returning either `False, some_innocent_default_value` or `True, computed value`. That is, in either case, only one piece of useful info is actually returned – `False` for the former, `computed_value` for the latter. I do suspect that exceptions over flags would be clearer; does the below seem a helpful pattern?

```
def oofp():
    first step
    if something_wrong:
        raise MajorProblemError(msg)
    next_step
    if something_wrong:
        raise SlightlyIllConditionedError(msg)
    final step
    return ans

class OOFP(object):
    def doing_something:
        try:
            ans = oofp()
        except SlightlyIllConditionedError as err:
            ans = massage_a_bit()
        # no need to handle MajorProblemError as execution ended
        next_step(ans)
```

psi-rking commented:
The short answer is that these routines get used in many different
contexts, and there are sometimes layers like torsion code calling bend
code.  Sometimes failure to be able to compute something is more severe,
sometimes less so.  I'm fairly sure you could accomplish the same thing
with exceptions; and I'm fairly sure the resulting code would be much
longer, harder to read and adapt.  Inability to intelligently deal with the
many reasons for 'False' is part of what made the C++ code so difficult to
improve.

In this particular case, failure to be able to compute an out-of-plane
angle that is in the internal coordinate set dictates a change in algorithm
is required.  Other times a 'check = false' return value might simply
indicate that this coordinate should not be included in the internal
coordinate set at all.

On Sat, Jan 13, 2018 at 5:30 PM, Lori A. Burns <notifications@github.com>
wrote:

> *@loriab* commented on this pull request.
> ------------------------------
>
> In Tutorials/13_Geometry-Optimization/opt_helper/oofp.py
> <https://github.com/psi4/psi4numpy/pull/37#discussion_r161383806>:
>
> > +    def qShowFactor(self):
> +        return 180.0 / np.pi
> +
> +    def qShow(self, geom):  # return in degrees
> +        return self.q(geom) * self.qShowFactor
> +
> +    @property
> +    def fShowFactor(self):
> +        return HARTREE2AJ * np.pi / 180.0
> +
> +    # compute angle and return value in radians
> +    def q(self, geom):
> +        check, tau = v3d.oofp(geom[self.A], geom[self.B], geom[self.C], geom[self.D])
> +        if not check:
> +            raise optExceptions.ALG_FAIL(
> +                "OOFP::compute.q: unable to compute out-of-plane value")
>
> It would be more usual for the fns in v3d that hit some tolerance that
> currently results in a check = False to instead raise an exception right
> when the error occurs in v3d, rather than passing the flag back up to here
> to handle. That way the calling function could proceed assuredly. Is there
> a reason to defer error handling back to the calling fn, like the above?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/psi4/psi4numpy/pull/37#pullrequestreview-88669592>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADguTOAWOahV64KqvFJM_VoX4u2ozT4Bks5tKTyigaJpZM4Qd2O2>
> .
>

psi-rking commented:
I believe the urgent, usability-related issues have been dealt with now.  I added commits that make the print statements python3 compatible, tested with python3.6, removed pubchem, and consolidated all imports in each file.

dgasmith commented:
Can you rebase this with master to see if we can get the tests to pass?

psi-rking commented:
My attempt at a rebase failed  (early at a commit time when many files were being moved around).
error: could not apply a492ee5... moved geometry optimization stuff into tutorials

Travis CI is showing the import of the helper files isn't working.

psi-rking commented:
Since the geometry tutorials are virtually isolated from the rest of the
code, if a rebase would really be helpful, I could do it manually.  Adding
all the changes to the current master.  I'm not sure this has anything to
do with the tests failings but would look cleaner.


On Wed, Jan 17, 2018 at 4:42 PM, Daniel Smith <notifications@github.com>
wrote:

> Can you rebase this with master to see if we can get the tests to pass?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/psi4/psi4numpy/pull/37#issuecomment-358474376>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADguTHfUXNy-BE8nOa4nLAVNvZvgreajks5tLndogaJpZM4Qd2O2>
> .
>

psi-rking commented:
If nobody objects for some reason, I would like to pull this pull request.  (Or is that pushing?)  I created a new one that is one clean commit.

dgasmith commented:
A rebase/merge is required to have the tests pass. It looks like you touched a number of files outside of your folder, please run a `git checkout master -- Tutorials/06_Molecular_Properties/CP-SCF.ipynb`. Where `master` needs to be changed to wherever the current `psi4/psi4numpy:master` HEAD currently resides. Its not typically a good idea to edit your `master` branch.

psi-rking commented:
Messages passed in transit.  I should have thought of trying that checkout
-- myself.
Don't know what I did before, but hopefully new branch is a clean add.
Then I guess there is still an import problem to solve.

On Thu, Jan 18, 2018 at 10:59 AM, Daniel Smith <notifications@github.com>
wrote:

> A rebase/merge is required to have the tests pass. It looks like you
> touched a number of files outside of your folder, please run a git
> checkout master -- Tutorials/06_Molecular_Properties/CP-SCF.ipynb. Where
> master needs to be changed to wherever the current psi4/psi4numpy:master
> HEAD currently resides. Its not typically a good idea to edit your master
> branch.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/psi4/psi4numpy/pull/37#issuecomment-358710895>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADguTGzg0QKtYlrf-1W16lHJfYX1f6XYks5tL3hXgaJpZM4Qd2O2>
> .
>

psi-rking commented:
New pull request is passing tests.

On Thu, Jan 18, 2018 at 11:01 AM, Rollin King <rking@bethel.edu> wrote:

> Messages passed in transit.  I should have thought of trying that checkout
> -- myself.
> Don't know what I did before, but hopefully new branch is a clean add.
> Then I guess there is still an import problem to solve.
>
>
> On Thu, Jan 18, 2018 at 10:59 AM, Daniel Smith <notifications@github.com>
> wrote:
>
>> A rebase/merge is required to have the tests pass. It looks like you
>> touched a number of files outside of your folder, please run a git
>> checkout master -- Tutorials/06_Molecular_Properties/CP-SCF.ipynb. Where
>> master needs to be changed to wherever the current psi4/psi4numpy:master
>> HEAD currently resides. Its not typically a good idea to edit your master
>> branch.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/psi4/psi4numpy/pull/37#issuecomment-358710895>, or mute
>> the thread
>> <https://github.com/notifications/unsubscribe-auth/ADguTGzg0QKtYlrf-1W16lHJfYX1f6XYks5tL3hXgaJpZM4Qd2O2>
>> .
>>
>
>

dgasmith commented:
Closed by #43.
